File: src/main/java/cube8540/oauth/authentication/credentials/oauth/token/application/ClientCredentialsTokenGranter.java
package cube8540.oauth.authentication.credentials.oauth.token.application;

import cube8540.oauth.authentication.credentials.oauth.client.domain.OAuth2ClientId;
import cube8540.oauth.authentication.credentials.oauth.error.InvalidGrantException;
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2ClientDetails;
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2TokenRequest;
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2AccessTokenRepository;
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2AuthorizedAccessToken;
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2TokenIdGenerator;
import lombok.Setter;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.security.oauth2.core.AuthorizationGrantType;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;

@Service
public class ClientCredentialsTokenGranter extends AbstractOAuth2TokenGranter {

    @Setter
    private boolean allowedRefreshToken = false;

    public ClientCredentialsTokenGranter(@Qualifier("defaultTokenIdGenerator") OAuth2TokenIdGenerator tokenIdGenerator,
                                         OAuth2AccessTokenRepository tokenRepository) {
        super(tokenIdGenerator, tokenRepository);
    }

    @Override
    public OAuth2AuthorizedAccessToken createAccessToken(OAuth2ClientDetails clientDetails, OAuth2TokenRequest tokenRequest) {
        if (!getTokenRequestValidator().validateScopes(clientDetails, tokenRequest.getScopes())) {
            throw InvalidGrantException.invalidScope("cannot grant scopes");
        }

        OAuth2AuthorizedAccessToken accessToken = OAuth2AuthorizedAccessToken.builder(getTokenIdGenerator())
                .client(new OAuth2ClientId(clientDetails.getClientId()))
                .expiration(extractTokenExpiration(clientDetails))
                .tokenGrantType(AuthorizationGrantType.CLIENT_CREDENTIALS)
                .scopes(extractGrantScope(clientDetails, tokenRequest))
                .issuedAt(LocalDateTime.now(clock))
                .build();
        if (allowedRefreshToken) {
            accessToken.generateRefreshToken(refreshTokenGenerator(), extractRefreshTokenExpiration(clientDetails));
        }
        accessToken.generateComposeUniqueKey(getComposeUniqueKeyGenerator());
        return accessToken;
    }
}

File: src/main/kotlin/cube8540/oauth/authentication/oauth/token/application/ClientCredentialsTokenGranter.kt
package cube8540.oauth.authentication.oauth.token.application

import cube8540.oauth.authentication.oauth.client.domain.OAuth2ClientId
import cube8540.oauth.authentication.oauth.error.InvalidGrantException
import cube8540.oauth.authentication.oauth.security.OAuth2ClientDetails
import cube8540.oauth.authentication.oauth.security.OAuth2TokenRequest
import cube8540.oauth.authentication.oauth.token.domain.OAuth2AccessTokenRepository
import cube8540.oauth.authentication.oauth.token.domain.OAuth2AuthorizedAccessToken
import cube8540.oauth.authentication.oauth.token.domain.OAuth2TokenIdGenerator
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.security.oauth2.core.AuthorizationGrantType
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class ClientCredentialsTokenGranter @Autowired constructor(
    @Qualifier("defaultTokenIdGenerator")
    tokenIdGenerator: OAuth2TokenIdGenerator,

    tokenRepository: OAuth2AccessTokenRepository
): AbstractOAuth2TokenGranter(tokenIdGenerator, tokenRepository) {

    var allowedRefreshToken: Boolean = false

    override fun createAccessToken(clientDetails: OAuth2ClientDetails, tokenRequest: OAuth2TokenRequest): OAuth2AuthorizedAccessToken {
        if (!tokenRequestValidator.validateScopes(clientDetails, tokenRequest.scopes)) {
            throw InvalidGrantException.invalidScope("cannot grant scope")
        }

        val accessToken = makeNewToken(clientDetails, tokenRequest)
        if (allowedRefreshToken) {
            accessToken.generateRefreshToken(refreshTokenGenerator(), extractRefreshTokenExpiration(clientDetails))
        }
        return accessToken
    }

    private fun makeNewToken(clientDetails: OAuth2ClientDetails, tokenRequest: OAuth2TokenRequest) = OAuth2AuthorizedAccessToken(
        tokenIdGenerator = tokenIdGenerator,
        username = null,
        client = OAuth2ClientId(clientDetails.clientId),
        scopes = extractGrantScope(clientDetails, tokenRequest).toMutableSet(),
        expiration = extractTokenExpiration(clientDetails),
        tokenGrantType = AuthorizationGrantType.CLIENT_CREDENTIALS,
        issuedAt = LocalDateTime.now(clock)
    )
}